### PR TITLE
Support SHA3 384/512

### PR DESCRIPTION
**Pull request by justsmth**
### Issues:
N/A

### Description of changes: 
* Adds support for SHA3-384 and SHA3-512.
* Uses NIST test vectors

### Call-outs:
N/A

### Testing:
All tests pass.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

File: aws-lc-rs/src/digest/sha.rs [hunk 1]
```diff
@@ -143,3 +193,17 @@ fn sha512_256_digest(msg: &[u8], output: &mut [u8]) {
...
+    let digest = ctx.finish();
+    output[0..SHA3_384_OUTPUT_LEN].copy_from_slice(digest.as_ref());
+}
+
+fn sha3_512_digest(msg: &[u8], output: &mut [u8]) {
+    let mut ctx = Context::new(&SHA3_512);
+    ctx.update(msg);
+    let digest = ctx.finish();
+    output[0..SHA3_512_OUTPUT_LEN].copy_from_slice(digest.as_ref());
+}
```
Comment (justsmth) [inline aws-lc-rs/src/digest/sha.rs:209]: I couldn't find a "one-shot" digest function for SHA3 in AWS-LC.

Comment (samuel40791765) [reply aws-lc-rs/src/digest/sha.rs:209]: Yeah I looked around and there doesn't seem to be a "one-shot" SHA3 digest function in OpenSSL either.

File: aws-lc-rs/src/digest/sha.rs [hunk 2]
```diff
@@ -143,3 +193,17 @@ fn sha512_256_digest(msg: &[u8], output: &mut [u8]) {
         aws_lc::SHA512_256(msg.as_ptr(), msg.len(), output.as_mut_ptr());
     }
 }
+
+fn sha3_384_digest(msg: &[u8], output: &mut [u8]) {
+    let mut ctx = Context::new(&SHA3_384);
+    ctx.update(msg);
+    let digest = ctx.finish();
+    output[0..SHA3_384_OUTPUT_LEN].copy_from_slice(digest.as_ref());
```
Comment (dkostic) [inline aws-lc-rs/src/digest/sha.rs:201]: I assume:
```
Context::new(&SHA3_384) calls EVP_DigestInit,
ctx.update(msg) calls EVP_DigestUpdate,
digest = ctx.finish() calls EVP_DigestFinal.
```
right?

File: aws-lc-rs/tests/digest_test.rs [hunk 3]
```diff
@@ -152,10 +183,54 @@ mod digest_shavs {
         assert_eq!(expected_count, 100);
     }
 
-    shavs_tests!(SHA1, SHA1_FOR_LEGACY_USE_ONLY);
-    shavs_tests!(SHA256, SHA256);
-    shavs_tests!(SHA384, SHA384);
-    shavs_tests!(SHA512, SHA512);
+    #[allow(clippy::cast_sign_loss)]
+    fn run_monte_carlo_fips_200_test(
```
Comment (samuel40791765) [inline aws-lc-rs/tests/digest_test.rs:187]: So FIPS 180-4 is the original Hash standard and FIPS 202 is for SHA3.. Should this be `run_monte_carlo_fips_202_test` or are these test vectors from the FIPS 200 publication?

**Review by samuel40791765 — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Review by samuel40791765 — DISMISSED** [review]

File: aws-lc-rs/src/digest/sha.rs (hunk 2 above)
Comment (justsmth) [reply aws-lc-rs/src/digest/sha.rs:201]: Right. To be specific:
* `Context::new(&SHA3_384)` calls [`EVP_MD_CTX_init` and `EVP_DigestInit_ex`](https://github.com/awslabs/aws-lc-rs/blob/main/aws-lc-rs/src/digest/digest_ctx.rs#L17-L18).
* `ctx.update(msg)` calls [`EVP_DigestUpdate`](https://github.com/awslabs/aws-lc-rs/blob/main/aws-lc-rs/src/digest.rs#L105)
* `ctx.finish()` calls [`EVP_DigestFinal`](https://github.com/awslabs/aws-lc-rs/blob/main/aws-lc-rs/src/digest.rs#L131)
* When the value goes out of scope, [`EVP_MD_CTX_cleanup`](https://github.com/awslabs/aws-lc-rs/blob/main/aws-lc-rs/src/digest/digest_ctx.rs#L46) is called.

**Review by samuel40791765 — APPROVED** [review]

File: aws-lc-rs/tests/digest_test.rs (hunk 3 above)
Comment (justsmth) [reply aws-lc-rs/tests/digest_test.rs]: DOH. It was a typo. Fixed in next commit.

**Pull request merged**